test/integration_test.py

```diff
@@ -724,8 +728,13 @@ def run(thread_count):
                 }
                 self.portal_service._crud(lambda db: [*db, mock_entry])
 
-        with ThreadPoolExecutor(max_workers=n_threads) as executor:
+        executor = ThreadPoolExecutor(max_workers=n_threads)
+        try:
             futures = [executor.submit(run, i) for i in range(n_tasks)]
+            executor.shutdown(wait=True)
+        finally:
+            running = False
+            executor.shutdown(wait=True)
```

Comment:
I'm a little confused why this is called twice, in line 734, it looks like the shutdown is used to ensure that no additional futures are able to execute.

Why would we need a subsequent call?

Reply:
To prevent the following circumstance:
1. Exception is raised during the first call to `execute`, before the worker threads are finished.
2. `running` is set to False. 
3. The `finally` block exits.
4. Worker threads are stuck waiting for the network response, don't yet reach the `return` statement.
5. `tearDown` runs.
6. Worker threads finish working, _after_ `teardown` has run, thus invalidating this fix.
7. Worker threads reach the `return` statement.

The subsequent call to `shutdown` enforces the following sequence, as long as no further exceptions are raised:
1. Exception is raised during the first call to `execute`, before the worker threads are finished.
2. `running` is set to False. 
3. Worker threads are stuck waiting for the network response, don't yet reach the `return` statement.
4. `shutdown` blocks until the threads are finished.
5. Worker threads reach the `return` statement.
6. The `finally` block exits.
7. `tearDown` runs.